### PR TITLE
Directives handling for #47, #48, #49, #50

### PR DESCRIPTION
**Pull request by epii-1**
Drafting this one early in case there are any comments on the implementation, to be able to handle these early. 

It's still a work in progress, all changes to the generator of actually pushing directives from the original schema, to the api-schema is still missing etc..

Ignore the white space changes, VS went dumbmode when I copied code between files, and I will try to revert those changes later...

The interesting part for now, is the function at the very button; `addPostCodeDirectivesForTypes`. This function will be responsible for enforcing all directives, at the very end of any transaction.  This will allow the user to do some very strange things during the actual transaction, as long as the results are consistent. I.e. a user may temporary create something that breaks the `@noloops` directive, as long as they change it to not break it later, during the same transaction. It may not really make sense now, but it may become useful once #52 is done.
`ctxt.trans.postCode` is, just like `.code`, a simple string array, that is appended at the very end of `.code`, just before the transaction is executed.

Using this approach we do not have to push the directives to any input fields, as suggested in e.g. https://github.com/LiUGraphQL/woo.sh/issues/47#issuecomment-627230501 (which I would otherwise have agreed with). Instead we can keep the directives at the fields they are defined on.

This might not be the simplest approach, checking the inputs of a mutation is probably easier, but this approach should never fail, and it gives us/the user the greatest flexibility as we only enforce the directives at the very end.

@hartig

### CODE REVIEW

**Comment by hartig** [comment]
It is great to see that you are looking at the issue with the broader picture in mind! (..namely, that the constraints should not be checked on a per operation basis but, instead, at the end of the whole transaction.)

I have only two comments at this point:

1. You say that by "using this approach we do not have to push the directives to any input fields [...] Instead we can keep the directives at the fields they are defined on." Proposing to copy them to the input types in the API schema rather than to the object types was deliberate. From the perspective of an API, the constraints are relevant only for the mutations but not for the queries. Hence, for the object types in the API schema (which specify what can be asked for in queries), the constraints are irrelevant. In contrast, they are relevant for the input types that are used in the mutations. That's why I want the constraints-related directives from a type in the DB schema to be carried over to the corresponding input types in the API schema rather than to the corresponding object type.

2. A minor thing:  The name "postCode" is a bad choice for two reasons: First, it reminds me of the English word postcode (a short form for postal code). Second, this name does not explicitly capture what this code is about. I think a better name would be finalConstraintChecks.

**Comment by epii-1** [comment]
On your point 1:
While I in theory do agree with you  that it makes more sense to have the directives on the input types instead of the object types, it is easier in practice to keep them at the object types. Firstly our calls to `create`, `update` etc. contains `returnType` and the functions uses `returnType` for all type and interface checks anyway. Secondly we don't have direct access to the `_InputTo...` types, only direct access to the objects and implicite access to the input types. So we would have to interpret the `_InputTo...` type from `returnType` in any case. This becomes even more apparent for recursive calls to `create`. Thirdly we would have to have different interpreters for updates and creates, as the input types differ. Only using the `returnType` and keeping the directives to the object types hence seems neater in practice.
In short: we already break this practice by using the object type instead of the input types to handle everything.

**Comment by hartig** [comment]
I look at this question from the perspective of a user of GraphQL API generated by woo.sh. As such a user, I would find it strange that the input types which I use when writing my mutation request do not contain any indication of such constraints.

Your argument that we are already using the `returnType` in the implementation of `create`, `update`, etc., is also not totally convincing: It would be very easy to also pass the corresponding input type into these functions. The only thing that would be necessary to do that is to extend the template based on which 'resolvers.js' is generated. For instance, instead of generating the following resolver
```
createAudioFile: (parent, args, context, info) =>
    driver.create(true, context, args.data, info.schema.getType('AudioFile'), info),
```
we may generate a resolver that looks as follows:
```
createAudioFile: (parent, args, context, info) =>
    driver.create(true, context, args.data, info.schema.getType('AudioFile'), info.schema.getType('_InputToCreateAudioFile'), info),
```

**Comment by epii-1** [comment]
For me this now turned into two (connected but different) questions:

1. > I look at this question from the perspective of a user of GraphQL API generated by woo.sh. As such a user, I would find it strange that the input types which I use when writing my mutation request do not contain any indication of such constraints.

I.e. should the input types contain the directives? They indeed should and it's a good thing you pointed it out. I'd probably not have thought about it otherwise.
Should the object types still contain the (input relevant) directives? In some cases they of course have to be in the object type as well (specifically for `@requiredForTarget` currently). For the others I don't see any problem if they do. If anything it gives the user slightly more information about the object(s) they are querying

2. Should we pass the input types to the functions handling the mutations? We don't need to do it unless the object types does not contain the input relevant directives. It's not really an issue, but a matter of convenience.

**Comment by hartig** [comment]
I see your point, and I agree. Please proceed.

**Comment by epii-1** [comment]
I almost thought I had this one done. Then I reread our earlier discussion and realized that I forgot to push directives onto the input types.

@hartig if you have the time, can you take a quick peek at https://gitlab.com/spirit-tools/arangodb-graphql-server/-/commit/c4f1bd553b8bb78d5d0158fc554ab24500e6a1e3 and see if there is anything that should be different in the formatting of the api-schema.
My biggest question is probably if the `@keys` directives should still be part of the api-schema.
Note that I have already added changes solving https://gitlab.com/spirit-tools/utils/graphql-schema/-/issues/62 there (I.e. removed some `@requiredForTarget` directives), but my question about the schema in the discussion of that issue still stands.

**Comment by epii-1** [comment]
Continuing the discussion from before
```
I look at this question from the perspective of a user of GraphQL API generated by woo.sh. As such a user, I would find it strange that the input types which I use when writing my mutation request do not contain any indication of such constraints.
```
This actually turned a bit more complicated then I thought. For example `@required` can't be used on input fields, and hence we can't clearly show that this field is required for say a create mutation. 
For the four directives we define ourselves, we would need to mark those as usable on `INPUT_FIELD_DEFINITION` to be able to have them on the fields of the input types.
So do we still want to do this for all our directives, even though the "standard" ones can't be used in this way? Or do we even want to extend the "standard" directives to be usable on fields of the input types too?

**Comment by hartig** [comment]
Hi Leif, I am first responding to your remarks and questions in [your latest comment above](https://github.com/LiUGraphQL/woo.sh/pull/54#issuecomment-635877142).

> For example @required can't be used on input fields, and hence we can't clearly show that this field is required for say a create mutation.

I assume you mean input types rather than "input fields." Then, @required does not need to be carried over to the input types. Instead, in input types, adding the exclamation mark ("!") to the type of a field in an input type means that this is a mandatory field that has to be provided.

> For the four directives we define ourselves, we would need to mark those as usable on INPUT_FIELD_DEFINITION to be able to have them on the fields of the input types.

Good point. Yes. While this extended definition of these four directives should not be in the DB schema, it should be in the API schema. Hence, when carrying over the definitions of these four directives, they need to be extended with the INPUT_FIELD_DEFINITION option.

Similarly, the definition of the @key directive has to be extended in a similar way, assuming we want to carry over the given @key directives from the DB schema to the corresponding input types in the API schema.

> So do we still want to do this for all our directives, even though the "standard" ones can't be used in this way? Or do we even want to extend the "standard" directives to be usable on fields of the input types too?

Even if I am not sure what you mean by "standard" directives, my answer to the first of these two questions is probably yes.

> What about inputs for updates? 

All the directives that are carried over to the input types for create operations need to be carried over in the same way also to the input types for the update operations. In this context, note again that @required does not need to be carried over. Instead, this is captured by the exclamation marks, and if you look at the input types generated for the update operations, your will see that there are no exclamation marks.

> then we would need something similar for all directives to differ between object field directives and input field directives to clearly indicate what the directives actually mean.

Not sure what you mean here. Can you elaborate this a bit more?

**Comment by hartig** [comment]
Leif, in your other comment you had one more question:

> My biggest question is probably if the @keys directives should still be part of the api-schema.

Yes, I think it makes sense to carry over the @keys directives to the corresponding input types similar to the other of our constraint-related directives.

**Comment by epii-1** [comment]
> Even if I am not sure what you mean by "standard" directives, my answer to the first of these two questions is probably yes.
I just mean the four directives that are always available in graphql: `@required`, `@deprecated`, `@skip` and `@include`.

> > then we would need something similar for all directives to differ between object field directives and input field directives to clearly indicate what the directives actually mean.
> 
> Not sure what you mean here. Can you elaborate this a bit more?

Take the (half bad) example of the `@requiredForTarget` directive: if some `input _inputA` which is used to create objects of `type A` where both `A` and `_inputA` has a field marked with a `@requiredForTarget` directive. Does that mean that the target does actually require an edge from an object off `_inputA`, or does this directive just mark that the target requires an edge from an object off type `A`? It might be a bad example, as it's quite clear what we mean in this example. `@required` directives are actually even more confusing, but we have already said that they do not appear in input types, so that's already solved.

It might very well be a non-issue as of now, as non of the current directives (except `@required`, which is already handled by not using it for inputs) is open for misunderstanding.

**Comment by hartig** [comment]
Leif,

> I just mean the four directives that are always available in graphql: `@required`, `@deprecated`, `@skip` and `@include`.

Ah, right, the directives that are defined in the GraphQL spec. By the way, `@required` is not one of them: http://spec.graphql.org/June2018/#sec-Type-System.Directives

> Take the (half bad) example of the `@requiredForTarget` directive: if some `input _inputA` which is used to create objects of `type A` where both `A` and `_inputA` has a field marked with a `@requiredForTarget` directive. Does that mean that the target does actually require an edge from an object off `_inputA`, or does this directive just mark that the target requires an edge from an object off type `A`?

Our notions of constraints are defined only for DB schemas not for API schemas. Additionally, input types do not really have a presence in the DB. So, in this sense, I would not worry so much about this question.

**Comment by epii-1** [comment]
(How did I miss that we actually define `@required` ourselves?)

Olaf,

> Our notions of constraints are defined only for DB schemas not for API schemas. Additionally, input types do not really have a presence in the DB. So, in this sense, I would not worry so much about this question.

In the terms of actual implementation there is no issue here. But your earlier comment of

> I look at this question from the perspective of a user of GraphQL API generated by woo.sh

is what got me thinking about this. Is it clear for the user what the directive actually means when it's placed on a field in an input type?

**Comment by hartig** [comment]
> Is it clear for the user what the directive actually means when it's placed on a field in an input type?

I would say yes.

**Comment by epii-1** [comment]
Good, then I wont complicate things any further.

**Comment by hartig** [comment]
No worries. It is good to have these questions asked ;-)

**Comment by epii-1** [comment]
I'm marking this as ready for review (and hence for pending merge) now.

To see actual output format (in the SPIRIT case), see https://gitlab.com/spirit-tools/mediator-service/-/commit/c4f1bd553b8bb78d5d0158fc554ab24500e6a1e3. Note that we currently push basically all directives to both inheriting types, as well as input types.

Issue #55 should be solved by this one (as I had to do some changed to that part anyway), issue #47, #48, #49 and #50 are all partially solved. As some functions are missing (update-edge operations for example), I can not claim that they are completely solved. But adding calls to `addfinalDirectiveChecksForType` when they are implemented should be enough.

`client_tests.js` is now stating in the description that it does not handle directives and `process.env.DISABLE_DIRECTIVES_CHECKING` have been introduced to disable directives checks in `driver.js`.

I guess we should start with Olaf quickly checking the formatting of the API schema before someone takes the time to do a full review.

@hartig @keski @rcapshaw

**Comment by hartig** [comment]
Leif, all,

Leif, thanks for your efforts so far. I have looked at the generated API schema and noticed that some things are missing:
1. Directives are not copied to the relevant input types. For instance, in the DB schema, the type field `locations` in the type `TimelineRegionsInVideoFile` has both the `@uniqueForTarget` directive and the `@requestForTarget` directive. These directives have been carried over to `TimelineRegionsInVideoFile` in the API schema, as we have agreed in our discussion above. However, these directives have not been copied to the corresponding input types, `_InputToCreateTimelineRegionsInVideoFile` and `_InputToUpdateTimelineRegionsInVideoFile`.
2. Similarly, the new automatically added directives `@_requiredForTarget_AccordingToInterface` and `@_uniqueForTarget_AccordingToInterface` are not at the corresponding input types.
3. The `@key` directives should also be copied to the corresponding input types (but they are not at the moment).
4. The declaration of these directives in the API schema is not extended with the `INPUT_FIELD_DEFINITION` option as discussed above.

**Comment by epii-1** [comment]
@hartig 
1-2. Good catch. I actually thought I checked these. Apparently not.
3-4. I actually linked an old commit by mistake. These two where actually already fixed.
Should have linked https://gitlab.com/spirit-tools/mediator-service/-/compare/master...directives?w=1 instead (to see all changes), or https://gitlab.com/spirit-tools/mediator-service/-/commit/c72bca6edfb68f347bde4823e75c8b2aacc828e4 to see that 1-2 is now solved.

I was now able to remove the strange `on ARGUMENT_DEFINITION` I for some reason needed before. Was never able to pinpoint why they were needed, as they should not have been needed.

**Comment by hartig** [comment]
Thanks Leif. I have checked again and confirm that now this looks good in terms of the API schema.

**Comment by keski** [comment]
As far as I can tell, the schema generator works as intended. However, checking the functionality with respect to the ArangoDB driver is not so straightforward. I think we need to set up individual tests for these. Still I think we can merge this feature branch into master at this point, and open a separate issue for the tests, since there's an option to disable the directive checks.

**Comment by hartig** [comment]
Sounds good. Please set the default value of the config option for the constraint checks to _disabled_ for the moment and, after that, merge.
